source/val/validate_adjacency.cpp

```diff
@@ -33,13 +33,14 @@ spv_result_t ValidateAdjacency(ValidationState_t& _, size_t idx) {
 
   switch (inst.opcode()) {
     case SpvOpPhi:
+      while (idx > 0 && instructions[idx - 1].opcode() == SpvOpLine) --idx;
```

Comment:
I feel like this whole pass would be better if it got pulled out of the instruction loop in validate and moved below. Then you could just flag state instead of doing the same searches over and over again.

Reply:
Do you mean adding a function like

```
spv_result_t ValidateAdjacencyForOpPhi(ValidationState_t& _) {
  const auto& instructions = _.ordered_instructions();
  bool valid_to_place = false;

  for (size_t i = 0; i < instructions.size(); ++i) {
    const auto& inst = instructions[i];
    switch (inst.opcode()) {
      case SpvOpLabel:
        valid_to_place = true;
        break;
      case SpvOpPhi:
      case SpvOpLine:
        if (!valid_to_place) {
          return _.diag(SPV_ERROR_INVALID_DATA, &inst)
              << "OpPhi must appear before all non-OpPhi instructions "
              << "(except for OpLine, which can be mixed with OpPhi).";
        }
        break;
      default:
        valid_to_place = false;
        break;
    }
  }

  return SPV_SUCCESS;
}
```

and calling it right after the instruction loop in validate?

Reply:
Yes, something like that and calling right before ValidateEntryPoints(). Then we can use this structure for #1923 as well.

Reply:
Updated.